Library/OcAppleKernelLib/KextPatcher.c

```diff
@@ -84,33 +90,66 @@ PatcherInitContextFromBuffer (
   // and request PRELINK_KERNEL_IDENTIFIER.
   //
 
-  if (!MachoInitializeContext (&Context->MachContext, Buffer, BufferSize, 0)) {
+  if (!MachoInitializeContext (&Context->MachContext, Buffer, BufferSize, 0, Is32Bit)) {
     DEBUG ((DEBUG_INFO, "OCAK: Patcher init from buffer %p %u has unsupported mach-o\n", Buffer, BufferSize));
     return EFI_INVALID_PARAMETER;
   }
 
-  Segment = MachoGetSegmentByName64 (
-    &Context->MachContext,
-    "__TEXT"
-    );
-  if (Segment == NULL || Segment->VirtualAddress < Segment->FileOffset) {
-    return EFI_NOT_FOUND;
+  //
+  // 32-bit can be of type MH_OBJECT, which has all sections in a single unnamed segment.
+  // We'll fallback to that if there is no __TEXT segment.
+  //
+  if (Is32Bit) {
+    Segment32 = MachoGetSegmentByName32 (
+      &Context->MachContext,
+      "__TEXT"
+      );
+    if (Segment32 == NULL || Segment32->VirtualAddress < Segment32->FileOffset) {
+      Section32 = MachoGetSegmentSectionByName32 (
+        &Context->MachContext,
+        "",
+        "__text"
+        );
+      if (Section32 == NULL) {
+        return EFI_NOT_FOUND;
+      }
+      VirtualAddress = Section32->Address;
+      FileOffset     = Section32->Offset;
+    } else {
+      VirtualAddress = Segment32->VirtualAddress;
+      FileOffset     = Segment32->FileOffset;
+    }
+
+  } else {
+    Segment64 = MachoGetSegmentByName64 (
+      &Context->MachContext,
+      "__TEXT"
+      );
+    if (Segment64 == NULL || Segment64->VirtualAddress < Segment64->FileOffset) {
+      return EFI_NOT_FOUND;
+    }
+
+    VirtualAddress = Segment64->VirtualAddress;
+    FileOffset     = Segment64->FileOffset;
   }
 
-  Context->VirtualBase   = Segment->VirtualAddress - Segment->FileOffset;
-  Context->VirtualKmod   = 0;
-  Context->KxldState     = NULL;
-  Context->KxldStateSize = 0;
-
-  Status = InternalConnectExternalSymtab (
-    &Context->MachContext,
-    &InnerContext,
-    Buffer,
-    BufferSize,
-    NULL
-    );
-  if (EFI_ERROR (Status)) {
-    return Status;
+  Context->Is32Bit        = Is32Bit;
+  Context->VirtualBase    = VirtualAddress < FileOffset ? VirtualAddress - FileOffset : FileOffset; // FIXME: 32-bit MH_OBJECT
```

Comment:
Not sure I understand the FIXME. Maybe this value needs to be signed now or we need to rely on unsigned wraparound? `VirtualBase` is essentially the value we should subtract from a virtual pointer to translate a virtual address (for instance, an address of a symbol) to a file offset for patching.